### ISSUE TITLE
In-place Realization

### ISSUE THREAD

**Issue by WillianJunior** [halide/Halide#4613]
According to the documentation of Halide::Func::realize():

"If the buffer is also one of the arguments to the function, strange things may happen, as the pipeline isn't necessarily safe to run in-place."

As such, how unsafe realization on Halide can be? For example, should I implement a wavefront propagation algorithm, what could I expect from Halide?

**Comment by abadams** [comment]
There are situations you can get into, depending on how you write the schedule, where the same point in a Func gets evaluated multiple times. If the output buffer is actually just an alias of the input buffer, this can affect the results (e.g. consider out(x, y) = in(x, y) + 1). Producing different outputs depending on the schedule is a violation of the core language promise of Halide, so we call it unsafe.

One way to avoid it is by using GuardWithIf or RoundUp as the tail strategy on all splits so that no point ever gets redundantly re-evaluated. But there's also a way to write a pipeline that reads from its output, and in that case you could have a pipeline with just the output, and no inputs, and it would be safe. It looks like this (using an in-place integration as the example algorithm):

```
Func out;
Var x, y;
out(x, y) = undef<float>();
RDom r(1, 100)
out(r, y) = out(r - 1, y) + out(r, y);
out(x, r) = out(x, r - 1) + out(x, r);

```

Storing "undef" to a Func is a no-op (that first definition compiles to no code), so for Funcs representing outputs it leaves whatever value is already in the buffer in place. Then you can read from the buffer in the update definition. Undef is unfortunately overloaded, because for non-outputs it just gets you uninitialized memory.

**Comment by WillianJunior** [comment]
From what I understood, the only way for Halide to be able to produce unintended results is through in-place update functions (e.g., the out(x,y)=in(x,y)+1 case with the output as an alias of the input) with ShiftInwards tail strategy. Is that correct or could there be another unsafe alternative?

On a more concrete example, assume a wavefront propagation of any operation (e.g., max) on an anti-raster order on an input image buffer sized h X w, on which "ar" is realized:
```
Var x,y;
ar(x,y) = undef<uint8_t>();

RDom r(0, w, 0, h);
ar(r.x-w-1, r.y-h-1) = max(ar(r.x-w-1, r.y-h-1), 
                       max(ar(r.x-w,   r.y-h-1), 
                           ar(r.x-w-1, r.y-h)));
```
This pipeline is impossible to parallelize in any way given the dependency on both r.x and r.y (except though allow_race_conditions()). However, for my problem I can iterate through this pipeline several times until the output becomes stable. What I'm seeing is that for different schedules the stability point of the image can differ. For my pipeline (which uses another raster stage before the next iteration) the race condition is benign, thus it shouldn't have different stability points. Could this discrepancy on the outputs be caused by some optimization performed by Halide after the IR compilation?

**Comment by abadams** [comment]
There are probably other obscure ways for it to happen, but I'm pretty sure that's the big one you'll hit. 

By different stability point, do you mean that when iterated till convergence you get a different output image? There are two types of race condition in your code - there's a read-modify-write hazard, and then there's the issue that the whole thing is order-dependent. Changes in the order shouldn't change the stability point, but depending on how things shake out with the hazard an iterative algorithm could indeed make zero forward progress on an iteration, falsely appearing to be stable. You may need to add atomic(true) to the schedule to handle the hazard, but that may inject mutexes and be slow, because it's not an in-place max=.

E.g. imagine that all pixels lockstep in parallel do all the loads, and then they all do the stores. That's going to converge in one iteration to something different to doing the iteration serially, or doing the iterations in parallel but atomically.

btw it seems weird to me that your coordinates are all negative. Is that intended? Normally for anti-raster scans I'd write w - 1 - r.x.

**Comment by abadams** [comment]
Wait, I take back my example - that would make forward progress, just not much of it. Hrm.

**Comment by WillianJunior** [comment]
> By different stability point, do you mean that when iterated till convergence you get a different output image?

Exactly. Sorry for the bad terminology. Convergence is the right term. Also, yes, the final convergence is different for distinct schedules.

> btw it seems weird to me that your coordinates are all negative. Is that intended? Normally for anti-raster scans I'd write w - 1 - r.x.

Again, you are right, w-1-r.x is correct. Typo.

> There are two types of race condition in your code - there's a read-modify-write hazard, and then there's the issue that the whole thing is order-dependent.

The RMW hazard shouldn't be a problem since on a single raster or anti-raster passage (both compute_root()) each pixel is only updated once (or at least it should). By forcefully breaking the order-dependence it just means that we may need more iterations, but it should converge to the same point.

> ... depending on how things shake out with the hazard an iterative algorithm could indeed make zero forward progress on an iteration, falsely appearing to be stable.

As you said, it would make small forward progress, but just in case I had the pipeline to iterate at least 2-3 times after a convergence just to be sure, and always the first convergence would be the final one.

Another important detail is that, although the convergence points for different schedules are distinct among themselves, multiple runs with the same schedule always return the same result, even with different numbers of iterations (which was expected given the whole benign race condition).

Also, just another thing that was on the back-burner: by using clamp() could I encounter the reevaluation problem of ShiftInwards tail strategy? If so, is there a recommended way to deal with border resolution with update functions with only reduction domains?

P.S., As I wrote this I realized that the RMW hazard could be dangerous if a pixel would be updated more than once. Since I use clamp to bound access to neighbor pixels could this be the a multiple updates source? If such I could pad my input image and then iterate through [1:w-2]. What would be better given the Halide semantics, (for w being the width of the un-padded image) Rdom x(1,w) or updating the Buffer min=1 and max=w?

**Comment by ronghongbo** [comment]
Adam, we have a case where we need the output buffer to be an alias of the input buffer. How to specify that, though? Thanks!

**Comment by abadams** [comment]
See my first post in this thread. For short pipelines you can specify it by having an output only, and using undef for its pure definition to be able to use the values already in it.

For longer pipelines with multiple compute_roots you can just pass the same buffer for input and output and be careful with the schedule.

**Comment by popizdeh** [comment]
I'm a bit confused by your last comment Andrew, would that mean my generator has no `Input` parameters? How do I run the pipeline using RunGenMain and pre-populate the buffer with image data? It's all very confusing...
